bioimageio/spec/utils/build_spec.py

```diff
@@ -0,0 +1,246 @@
+    elif weight_type == 'pickle':
+        weights = raw_nodes.WeightsEntry(
+            source=weight_uri,
+            sha256=weight_hash
+        )
+        weights = {'pickle': weights}
+        language = 'python'
+        framework = 'scikit-learn'
+
+        source_hash = None
+    else:
+        raise ValueError(f"Invalid weight type {weight_type}, expect one of {weight_types}")
+
+    return weights, language, framework, source_hash
+
+
+def _get_input_tensor(test_in, name, step, min_shape, preprocessing):
+    shape = test_in.shape
+    if step is None:
+        assert min_shape is None
+        shape_description = shape
+    else:
+        shape_description = {
+            'min': shape if min_shape is None else min_shape,
+            'step': step
+        }
+
+    kwargs = {}
+    if preprocessing is not None:
+        kwargs['preprocessing'] = preprocessing
+
+    inputs = raw_nodes.InputTensor(
+        name='input' if name is None else name,
+        data_type=str(test_in.dtype),
+        axes='bczyx' if test_in.ndim == 5 else 'bcyx',
+        shape=shape_description,
+        **kwargs
+    )
+    return inputs
+
+
+def _get_output_tensor(test_out, name,
+                       reference_input, scale, offset,
+                       postprocessing, halo):
+    shape = test_out.shape
+    if reference_input is None:
+        assert scale is None
+        assert offset is None
+        shape_description = shape
+    else:
+        assert scale is not None
+        assert offset is not None
+        shape_description = {
+            'reference_input': reference_input,
+            'scale': scale,
+            'offset': offset
+        }
+
+    kwargs = {}
+    if postprocessing is not None:
+        kwargs['postprocessing'] = postprocessing
+    if halo is not None:
+        kwargs['halo'] = halo
+
+    outputs = raw_nodes.OutputTensor(
+        name='output' if name is None else name,
+        data_type=str(test_out.dtype),
+        axes='bczyx' if test_out.ndim == 5 else 'bcyx',
+        shape=shape_description,
+        **kwargs
+    )
+    return outputs
+
+
+# TODO can we pattern match to decide if we have a url or doi?
+def _build_cite(cite):
+    citation_list = [
+        raw_nodes.CiteEntry(text=k, url=v) for k, v in cite.items()
+    ]
+    return citation_list
+
+
+# NOTE does not support multiple input / output tensors yet
+# to implement this we should wait for 0.4.0, see also
+# https://github.com/bioimage-io/spec-bioimage-io/issues/70#issuecomment-825737433
+def build_spec(
+    # model specific required
+    source: str,
+    model_kwargs: Dict[str, Union[int, float, str]],
+    weight_uri: str,
+    test_inputs: List[str],
+    test_outputs: List[str],
+    # general required
+    name: str,
+    description: str,
+    authors: List[str],
+    tags: List[str],
+    license: str,
+    documentation: str,
+    covers: str,
+    dependencies: str,
+    cite: Optional[List[str]],
```

Comment:
looking at `_build_cite` it should probably be:
```suggestion
    cite: Dict[str, str],
```

Reply:
Yes, indeed. I had it with a default value before and forgot to remove the optional.